src/common/pkcs11_lib.c

```diff
@@ -1585,6 +1585,7 @@ cert_object_t **get_certificate_list(pkcs11_handle_t *h, int *ncerts)
       cert_template[3].pValue = NULL;
       rv = h->fl->C_GetAttributeValue(h->session, object, cert_template, 4);
       if (rv != CKR_OK) {
+        free(id_value);
```

Comment:
Thanks for one more fixed leak! However, wouldn't it be better to free the allocated resources at the _single, common_ exit point? Maybe we should rewrite the function in the style:

```
/* Initialization block */
id_value = NULL;
cert_value = NULL;
...

/* The body */
...
if ... {
    goto getlist_error;
}
if ... {
    goto getlist_error;
}
...

/* Normal exit: */
ret = h->certs;
exit:
    free(id_value);
    free(cert_value);
    return ret;

/* Exit on error: */
getlist_error:
    ret = NULL;
    goto exit;
```

Reply:
Updated